File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/data/CustomEndpointsTest.java
package com.kinvey.androidTest.store.data;

import android.content.Context;
import android.os.Looper;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.test.RenamingDelegatingContext;
import android.test.suitebuilder.annotation.SmallTest;

import com.google.api.client.json.GenericJson;
import com.kinvey.android.AsyncCustomEndpoints;
import com.kinvey.android.Client;
import com.kinvey.android.callback.KinveyListCallback;
import com.kinvey.android.model.User;
import com.kinvey.android.store.UserStore;
import com.kinvey.androidTest.TestManager;
import com.kinvey.androidTest.model.Person;
import com.kinvey.java.CustomEndpoints;
import com.kinvey.java.core.KinveyClientCallback;
import com.kinvey.java.core.KinveyJsonResponseException;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.List;
import java.util.concurrent.CountDownLatch;

import static junit.framework.Assert.assertTrue;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class CustomEndpointsTest {

    private static final String X2 = "x2";
    private static final String GET_PERSON = "getPerson";
    private static final String GET_PERSON_LIST = "getPersonList";
    private static final String NOT_EXIST_CUSTOM_ENDPOINT = "notExistCustomEndpoint";    
    private static final String ID = "id";    
    private static final String RESULT = "res";
    private Client client;
    
    @Before
    public void setUp() throws InterruptedException, IOException {
        Context mMockContext = new RenamingDelegatingContext(InstrumentationRegistry.getInstrumentation().getTargetContext(), "test_");
        client = new Client.Builder(mMockContext).build();
        final CountDownLatch latch = new CountDownLatch(1);
        if (!client.isUserLoggedIn()) {
            new Thread(new Runnable() {
                public void run() {
                    Looper.prepare();
                    try {
                        UserStore.login(TestManager.USERNAME, TestManager.PASSWORD, client, new KinveyClientCallback<User>() {
                            @Override
                            public void onSuccess(User result) {
                                assertNotNull(result);
                                latch.countDown();
                            }

                            @Override
                            public void onFailure(Throwable error) {
                                assertNull(error);
                                latch.countDown();
                            }
                        });
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                    Looper.loop();
                }
            }).start();
        } else {
            latch.countDown();
        }
        latch.await();
    }

    /*
    * CustomEndpoint = 'x2'
    * function onRequest(request, response, modules) {
    *     response.body = {RESULT:request.body.id*2};
    *     response.complete();
    * }
    */
    @Test
    public void testCustomEndpoints() throws IOException {
        CustomEndpoints endpoints = client.customEndpoints(GenericJson.class);
        GenericJson genericJson = new GenericJson();
        int i = 1;
        genericJson.set(ID, i);
        CustomEndpoints.CustomCommand command = endpoints.callEndpointBlocking(X2, genericJson);
        GenericJson response = (GenericJson) command.execute();
        assertEquals(i * 2, ((BigDecimal) response.get(RESULT)).intValueExact());
    }

    @Test
    public void testCustomEndpointsDeprecatedMethod() throws IOException {
        CustomEndpoints endpoints = client.customEndpoints();
        GenericJson genericJson = new GenericJson();
        int i = 1;
        genericJson.set(ID, i);
        CustomEndpoints.CustomCommand command = endpoints.callEndpointBlocking(X2, genericJson);
        GenericJson response = (GenericJson) command.execute();
        assertEquals(i * 2, ((BigDecimal) response.get(RESULT)).intValueExact());
    }

    @Test
    public void testCustomEndpointsNotExist() throws IOException, InterruptedException {
        CustomEndpoints endpoints = client.customEndpoints(GenericJson.class);
        GenericJson genericJson = new GenericJson();
        CustomEndpoints.CustomCommand command = endpoints.callEndpointBlocking(NOT_EXIST_CUSTOM_ENDPOINT, genericJson);
        try {
            command.execute();
        } catch (KinveyJsonResponseException exception) {
            assertNotNull(exception);
        }
    }

    /*
    * CustomEndpoint = 'x2'
    * function onRequest(request, response, modules) {
    *     response.body = {RESULT:request.body.id*2};
    *     response.complete();
    * }
    */
    @Test
    public void testAsyncCustomEndpoints() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(GenericJson.class);
        GenericJson genericJson = new GenericJson();
        final int i = 1;
        genericJson.set(ID, i);
        DefaultKinveyClientCallback<GenericJson> callback = callEndpoint(X2, endpoints, genericJson);
        assertEquals(i * 2, ((BigDecimal) callback.result.get(RESULT)).intValueExact());
    }

    @Test
    public void testAsyncCustomEndpointsDeprecatedMethod() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints();
        GenericJson genericJson = new GenericJson();
        int i = 1;
        genericJson.set(ID, i);
        DefaultKinveyClientCallback<GenericJson> callback = callEndpoint(X2, endpoints, genericJson);
        assertEquals(i * 2, ((BigDecimal) callback.result.get(RESULT)).intValueExact());
    }

    @Test
    public void testAsyncCustomEndpointsNotExist() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(GenericJson.class);
        GenericJson genericJson = new GenericJson();
        final int i = 1;
        genericJson.set(ID, i);
        DefaultKinveyClientCallback<GenericJson> callback = callEndpoint(NOT_EXIST_CUSTOM_ENDPOINT, endpoints, genericJson);
        assertNotNull(callback.error);
    }

    @Test
    public void testAsyncCustomEndpointsInputNotExist() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(GenericJson.class);
        DefaultKinveyClientCallback<GenericJson> callback = callEndpoint(X2, endpoints, null);
        assertNotNull(callback.result);
    }

    /**
     CustomEndpoint = 'getPerson'
     function onRequest(request, response, modules) {
         response.body = {"name":"TestName"};
         response.complete();
     }
     */
    @Test
    public void testCustomEndpointsCustomClass() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(Person.class);
        DefaultKinveyClientCallback<Person> callback = callCustomJsonEndpointAsync(GET_PERSON, endpoints, null);
        assertNotNull(callback.result);
        assertTrue(callback.result instanceof Person);
    }

    @Test
    public void testCustomEndpointsWrongCustomClass() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(Person[].class);
        DefaultKinveyClientCallback<Person> callback = callCustomJsonEndpointAsync(GET_PERSON, endpoints, null);
        assertNotNull(callback.error);
    }

    /**
     CustomEndpoint = 'getPersonList'
     function onRequest(request, response, modules) {
         response.body = [{"name":"TestName"}];
         response.complete();
     }
     */
    @Test
    public void testCustomEndpointsCustomClassArray() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(Person.class);
        DefaultKinveyListCallback<Person> callback = callCustomJsonEndpointListAsync(GET_PERSON_LIST, endpoints, null);
        assertNotNull(callback.result);
        assertTrue(callback.result instanceof List);
        assertTrue(callback.result.get(0) instanceof Person);
    }

    @Test
    public void testCustomEndpointsWrongCustomClassArray() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(Person[].class);
        DefaultKinveyListCallback<Person> callback = callCustomJsonEndpointListAsync(GET_PERSON_LIST, endpoints, null);
        assertNotNull(callback.error);
    }

    @Test
    public void testCustomEndpointsCustomClassArrayNotExist() throws IOException, InterruptedException {
        AsyncCustomEndpoints endpoints = client.customEndpoints(Person.class);
        DefaultKinveyListCallback<Person> callback = callCustomJsonEndpointListAsync(NOT_EXIST_CUSTOM_ENDPOINT, endpoints, null);
        assertNotNull(callback.error);
    }

    private DefaultKinveyClientCallback<GenericJson> callEndpoint(final String endPointName, final AsyncCustomEndpoints customEndpoints, final GenericJson json) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyClientCallback<GenericJson> callback = new DefaultKinveyClientCallback<>(latch);
        new Thread(new Runnable() {
            public void run() {
                Looper.prepare();
                customEndpoints.callEndpoint(endPointName, json, callback);
                Looper.loop();
            }
        }).start();
        latch.await();
        return callback;
    }

    private DefaultKinveyClientCallback<Person> callCustomJsonEndpointAsync(final String endPointName, final AsyncCustomEndpoints customEndpoints, final GenericJson json) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyClientCallback<Person> callback = new DefaultKinveyClientCallback<>(latch);
        new Thread(new Runnable() {
            public void run() {
                Looper.prepare();
                customEndpoints.callEndpoint(endPointName, json, callback);
                Looper.loop();
            }
        }).start();
        latch.await();
        return callback;
    }

    private DefaultKinveyListCallback<Person> callCustomJsonEndpointListAsync(final String endPointName, final AsyncCustomEndpoints customEndpoints, final GenericJson json) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyListCallback<Person> callback = new DefaultKinveyListCallback<>(latch);
        new Thread(new Runnable() {
            public void run() {
                Looper.prepare();
                customEndpoints.callEndpoint(endPointName, json, callback);
                Looper.loop();
            }
        }).start();
        latch.await();
        return callback;
    }

    private static class DefaultKinveyClientCallback<T> implements KinveyClientCallback<T> {

        private CountDownLatch latch;
        private T result;
        private Throwable error;

        private DefaultKinveyClientCallback(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onSuccess(T result) {
            this.result = result;
            finish();
        }

        @Override
        public void onFailure(Throwable error) {
            this.error = error;
            finish();
        }

        void finish() {
            latch.countDown();
        }
    }

    private static class DefaultKinveyListCallback<T> implements KinveyListCallback<T> {

        private CountDownLatch latch;
        private List<T> result;
        private Throwable error;

        private DefaultKinveyListCallback(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onSuccess(List<T> result) {
            this.result = result;
            finish();
        }

        @Override
        public void onFailure(Throwable error) {
            this.error = error;
            finish();
        }

        void finish() {
            latch.countDown();
        }
    }

    @After
    public void tearDown() {
        if (client.getKinveyHandlerThread() != null) {
            try {
                client.stopKinveyHandlerThread();
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }

}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/datastore/CustomEndpointsTest.kt
package com.kinvey.androidTest.store.datastore

import android.content.Context
import android.os.Looper
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.AndroidJUnit4
import com.google.api.client.json.GenericJson
import com.kinvey.android.AsyncCustomEndpoints
import com.kinvey.android.Client
import com.kinvey.android.Client.Builder
import com.kinvey.android.callback.KinveyListCallback
import com.kinvey.android.model.User
import com.kinvey.android.store.UserStore.Companion.login
import com.kinvey.androidTest.TestManager
import com.kinvey.androidTest.model.Person
import com.kinvey.java.AbstractClient
import com.kinvey.java.CustomEndpoints
import com.kinvey.java.CustomEndpoints.CustomCommand
import com.kinvey.java.core.KinveyClientCallback
import com.kinvey.java.core.KinveyJsonResponseException
import junit.framework.Assert.assertTrue
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException
import java.math.BigDecimal
import java.util.concurrent.CountDownLatch

@RunWith(AndroidJUnit4::class)
@SmallTest
class CustomEndpointsTest {
    private var client: Client<User>? = null
    @Before
    @Throws(InterruptedException::class, IOException::class)
    fun setUp() {
        val mMockContext: Context? = InstrumentationRegistry.getInstrumentation().targetContext
        client = Builder<User>(mMockContext).build()
        val latch = CountDownLatch(1)
        if (client?.isUserLoggedIn == false) {
            Thread(Runnable {
                Looper.prepare()
                try {
                    login<User>(TestManager.USERNAME, TestManager.PASSWORD, client as AbstractClient<User>, object : KinveyClientCallback<User> {
                        override fun onSuccess(result: User?) {
                            assertNotNull(result)
                            latch.countDown()
                        }

                        override fun onFailure(error: Throwable?) {
                            assertNull(error)
                            latch.countDown()
                        }
                    })
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                Looper.loop()
            }).start()
        } else {
            latch.countDown()
        }
        latch.await()
    }

    /*
    * CustomEndpoint = 'x2'
    * function onRequest(request, response, modules) {
    *     response.body = {RESULT:request.body.id*2};
    *     response.complete();
    * }
    */
    @Test
    @Throws(IOException::class)
    fun testCustomEndpoints() {
        val endpoints: CustomEndpoints<GenericJson, GenericJson>? = client?.customEndpoints(GenericJson::class.java)
        val genericJson = GenericJson()
        val i = 1
        genericJson.set(ID, i)
        val command = endpoints?.callEndpointBlocking(X2, genericJson)
        val response = command?.execute()
        assertEquals(i * 2.toLong(), (response?.get(RESULT) as BigDecimal).intValueExact().toLong())
    }

    @Test
    @Throws(IOException::class)
    fun testCustomEndpointsDeprecatedMethod() {
        val endpoints: CustomEndpoints<GenericJson, GenericJson>? = client?.customEndpoints()
        val genericJson = GenericJson()
        val i = 1
        genericJson.set(ID, i)
        val command = endpoints?.callEndpointBlocking(X2, genericJson)
        val response = command?.execute()
        assertEquals(i * 2.toLong(), (response?.get(RESULT) as BigDecimal).intValueExact().toLong())
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testCustomEndpointsNotExist() {
        val endpoints: CustomEndpoints<GenericJson, GenericJson>? = client?.customEndpoints(GenericJson::class.java)
        val genericJson = GenericJson()
        val command = endpoints?.callEndpointBlocking(NOT_EXIST_CUSTOM_ENDPOINT, genericJson)
        try {
            command?.execute()
        } catch (exception: KinveyJsonResponseException) {
            assertNotNull(exception)
        }
    }

    /*
    * CustomEndpoint = 'x2'
    * function onRequest(request, response, modules) {
    *     response.body = {RESULT:request.body.id*2};
    *     response.complete();
    * }
    */
    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testAsyncCustomEndpoints() {
        val endpoints: CustomEndpoints<GenericJson, GenericJson>? = client?.customEndpoints(GenericJson::class.java)
        val genericJson = GenericJson()
        val i = 1
        genericJson.set(ID, i)
        val callback = callEndpoint(X2, endpoints as AsyncCustomEndpoints<GenericJson, GenericJson>, genericJson)
        assertEquals(i * 2.toLong(), (callback.result!![RESULT] as BigDecimal).intValueExact().toLong())
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testAsyncCustomEndpointsDeprecatedMethod() {
        val endpoints: CustomEndpoints<GenericJson, GenericJson>? = client?.customEndpoints()
        val genericJson = GenericJson()
        val i = 1
        genericJson.set(ID, i)
        val callback = callEndpoint(X2, endpoints as AsyncCustomEndpoints<GenericJson, GenericJson>, genericJson)
        assertEquals(i * 2.toLong(), (callback.result!![RESULT] as BigDecimal).intValueExact().toLong())
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testAsyncCustomEndpointsNotExist() {
        val endpoints: CustomEndpoints<GenericJson, GenericJson>? = client?.customEndpoints(GenericJson::class.java)
        val genericJson = GenericJson()
        val i = 1
        genericJson.set(ID, i)
        val callback = callEndpoint(NOT_EXIST_CUSTOM_ENDPOINT, endpoints as AsyncCustomEndpoints<GenericJson, GenericJson>, genericJson)
        assertNotNull(callback.error)
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testAsyncCustomEndpointsInputNotExist() {
        val endpoints: AsyncCustomEndpoints<GenericJson, GenericJson>? = client?.customEndpoints(GenericJson::class.java)
        val callback = callEndpoint(X2, endpoints as AsyncCustomEndpoints<GenericJson, GenericJson>, null)
        assertNotNull(callback.result)
    }

    /**
     * CustomEndpoint = 'getPerson'
     * function onRequest(request, response, modules) {
     * response.body = {"name":"TestName"};
     * response.complete();
     * }
     */
    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testCustomEndpointsCustomClass() {
        val endpoints: AsyncCustomEndpoints<GenericJson, Person>? = client?.customEndpoints(Person::class.java)
        val callback = callCustomJsonEndpointAsync(GET_PERSON, endpoints as AsyncCustomEndpoints<GenericJson, Person>?, null)
        assertNotNull(callback.result)
        assertTrue(callback.result is Person)
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testCustomEndpointsWrongCustomClass() {
        val endpoints: AsyncCustomEndpoints<GenericJson, Array<Person>>? = client?.customEndpoints(Array<Person>::class.java)
        val callback = callCustomJsonEndpointAsync(GET_PERSON, endpoints as AsyncCustomEndpoints<GenericJson, Person>?, null)
        assertNotNull(callback.error)
    }

    /**
     * CustomEndpoint = 'getPersonList'
     * function onRequest(request, response, modules) {
     * response.body = [{"name":"TestName"}];
     * response.complete();
     * }
     */
    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testCustomEndpointsCustomClassArray() {
        val endpoints: AsyncCustomEndpoints<GenericJson, Person>? = client?.customEndpoints(Person::class.java)
        val callback = callCustomJsonEndpointListAsync(GET_PERSON_LIST, endpoints as AsyncCustomEndpoints<GenericJson, Person>, null)
        assertNotNull(callback.result)
        assertTrue(callback.result is List<*>)
        assertTrue(callback.result?.get(0) is Person)
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testCustomEndpointsWrongCustomClassArray() {
        val endpoints: AsyncCustomEndpoints<GenericJson, *>? = client?.customEndpoints(Array<Person>::class.java)
        val callback = callCustomJsonEndpointListAsync(GET_PERSON_LIST, endpoints as AsyncCustomEndpoints<GenericJson, Person>?, null)
        assertNotNull(callback.error)
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testCustomEndpointsCustomClassArrayNotExist() {
        val endpoints: AsyncCustomEndpoints<GenericJson, Person>? = client?.customEndpoints(Person::class.java)
        val callback = callCustomJsonEndpointListAsync(NOT_EXIST_CUSTOM_ENDPOINT, endpoints, null)
        assertNotNull(callback.error)
    }

    @Throws(InterruptedException::class)
    private fun callEndpoint(endPointName: String, customEndpoints: AsyncCustomEndpoints<GenericJson, GenericJson>?, json: GenericJson?): DefaultKinveyClientCallback<GenericJson> {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyClientCallback<GenericJson>(latch)
        Thread(Runnable {
            Looper.prepare()
            customEndpoints?.callEndpoint(endPointName, json, callback)
            Looper.loop()
        }).start()
        latch.await()
        return callback
    }

    @Throws(InterruptedException::class)
    private fun callCustomJsonEndpointAsync(endPointName: String, customEndpoints: AsyncCustomEndpoints<GenericJson, Person>?, json: GenericJson?): DefaultKinveyClientCallback<Person> {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyClientCallback<Person>(latch)
        Thread(Runnable {
            Looper.prepare()
            customEndpoints?.callEndpoint(endPointName, json, callback)
            Looper.loop()
        }).start()
        latch.await()
        return callback
    }

    @Throws(InterruptedException::class)
    private fun callCustomJsonEndpointListAsync(endPointName: String, customEndpoints: AsyncCustomEndpoints<GenericJson, Person>?, json: GenericJson?): DefaultKinveyListCallback<Person> {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyListCallback<Person>(latch)
        Thread(Runnable {
            Looper.prepare()
            customEndpoints?.callEndpoint(endPointName, json, callback)
            Looper.loop()
        }).start()
        latch.await()
        return callback
    }

    private class DefaultKinveyClientCallback<T>(private val latch: CountDownLatch) : KinveyClientCallback<T> {
        var result: T? = null
        var error: Throwable? = null
        override fun onSuccess(result: T?) {
            this.result = result
            finish()
        }

        override fun onFailure(error: Throwable?) {
            this.error = error
            finish()
        }

        internal fun finish() {
            latch.countDown()
        }

    }

    private class DefaultKinveyListCallback<T>(private val latch: CountDownLatch) : KinveyListCallback<T> {
        var result: List<*>? = null
        var error: Throwable? = null
        override fun onFailure(error: Throwable?) {
            this.error = error
            finish()
        }

        internal fun finish() {
            latch.countDown()
        }

        override fun onSuccess(result: List<T>?) {
            this.result = result
            finish()
        }

    }

    @After
    fun tearDown() {
        if (Client.kinveyHandlerThread != null) {
            try {
                client?.stopKinveyHandlerThread()
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        }
    }

    companion object {
        private const val X2 = "x2"
        private const val GET_PERSON = "getPerson"
        private const val GET_PERSON_LIST = "getPersonList"
        private const val NOT_EXIST_CUSTOM_ENDPOINT = "notExistCustomEndpoint"
        private const val ID = "id"
        private const val RESULT = "res"
    }
}